fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSession.java

```diff
@@ -0,0 +1,173 @@
+ */
+package org.fcrepo.persistence.ocfl.impl;
+
+import static org.fcrepo.persistence.ocfl.api.CommitOption.NEW_VERSION;
+import static org.fcrepo.persistence.ocfl.api.CommitOption.TO_MUTABLE_HEAD;
+
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.fcrepo.persistence.ocfl.api.CommitOption;
+import org.fcrepo.persistence.ocfl.api.OCFLObjectSession;
+
+import edu.wisc.library.ocfl.api.OcflRepository;
+import edu.wisc.library.ocfl.api.model.ObjectId;
+
+/**
+ * Default implementation of an OCFL object session, which stages changes to the
+ * file system prior to committing.
+ *
+ * @author bbpennel
+ */
+public class DefaultOCFLObjectSession implements OCFLObjectSession {
+
+    private String objectIdentifier;
+
+    // Path where changes to the OCFL object in this session are staged
+    private Path stagingPath;
+
+    private Set<String> deletePaths;
+
+    private OcflRepository ocflRepository;
+
+    /**
+     * Instantiate an OCFL object session
+     *
+     * @param objectIdentifier identifier for the OCFL object
+     * @param stagingPath path in which changes to the OCFL object will be staged.
+     * @param ocflRepository the OCFL repository in which the object is stored.
+     */
+    public DefaultOCFLObjectSession(final String objectIdentifier, final Path stagingPath,
+            final OcflRepository ocflRepository) {
+        this.objectIdentifier = objectIdentifier;
+        this.stagingPath = stagingPath;
+        this.ocflRepository = ocflRepository;
+    }
+
+    private Set<String> getDeletePaths() {
+        if (deletePaths == null) {
+            // TODO check to see if a delete log exists, in case resuming after reboot
+            deletePaths = new HashSet<>();
+        }
+        return deletePaths;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void write(final String subpath, final InputStream stream) {
+        // TODO write contents to the subpath within the staging path
+
+        // Determine the staging path for the incoming content
+        resolveStagedPath(subpath);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * If the file was not newly created within this session, then the
+     * deletion will be recorded for replay at commit time.
+     */
+    @Override
+    public void delete(final String subpath) {
+        if (newInSession(subpath)) {
+            // TODO delete the file from the staging path
+        } else {
+            getDeletePaths().add(subpath);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    private boolean newInSession(final String subpath) {
+        // TODO determine if this subpath exists in the OCFL object
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public InputStream read(final String subpath) {
+        if (hasStagedVersion(subpath)) {
```

Comment:
is this staging or mutable head?

Reply:
Staging for the current session. Hopefully mutable head should be addressable via the ocfl client

Reply:
Mostly just double checking as there is a staging area as well as mutable head, but I suspect this is the mutable head staged version area.

Reply:
this is the stuff that has changed in the current session and hasn't been committed to ocfl yet (as either a version or mutant head). We still need to provide access to these staged files within the scope of the transaction

Reply:
I'm wondering if this should be called hasStagingVersion?  Picky, I know, but I'm stuck on the word "Version".

Reply:
Oh, I like your "hasStagedChanges"... I'm going to rename the ocfl client apis.